File: 剑指offer/25.合并两个排序的链表.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        new_head = ListNode(0)
        r = new_head

        p = l1
        q = l2

        while p and q:
            while p and q and p.val <= q.val:
                next = p.next

                p.next = r.next
                r.next = p

                r = p
                p = next

            while p and q and q.val <= p.val:
                next = q.next

                q.next = r.next
                r.next = q

                r = q
                q = next

        if q:
            r.next = q

        if p:
            r.next = p

        return new_head.next

File: 剑指offer/25.合并两个排序的链表.cpp
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x)
        : val(x)
        , next(NULL)
    {
    }
};

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2)
    {
        ListNode* p1 = l1;
        ListNode* p2 = l2;

        ListNode* new_head = new ListNode(0);
        ListNode* p = new_head;

        while (p1 && p2) {
            if (p1->val < p2->val) {
                p->next = p1;
                p1 = p1->next;
                p = p->next;
            } else {
                p->next = p2;
                p2 = p2->next;
                p = p->next;
            }
        }

        if (p1) {
            p->next = p1;
        }

        if (p2) {
            p->next = p2;
        }

        return new_head->next;
    }
};